### PR TITLE
Simplify changing Edition types

### PR DESCRIPTION
**Pull request by jamiecobbett**
The old code had lots of hard-coded bespoke conditions that examined the types
of the objects to decide what to copy. This changes it to use duck typing to
figure out what data can be copied between types.

This approach is shorter, simpler and handles more conversions and will handle
new types without them having to be explicitly added. I think it also reveals the
intent better.

The only downside I could think of is if there are fields which are in common between two formats would be duplicated in the fields _and_ the new body, but that seems better than losing them because they haven't been explicitly handled one-by-one.

### CODE REVIEW

**Comment by boffbowsh** [comment]
:+1: for intent, @binaryberry knows this area though so will let her provide proper approval

File: app/models/edition.rb [hunk 1]
```diff
@@ -203,26 +203,25 @@ def build_clone(target_class=nil)
...
+    # Where the type is not changing, the body will already have been copied
+    # above.
+    if target_class != self.class
+
+      # Copying of parts between types with parts is handled by Parted
+      if new_edition.is_a?(Parted) and !self.is_a?(Parted)
+        new_edition.parts.build(title: "Part One", body: whole_body, slug: "part-one")
+      elsif new_edition.is_a?(TransactionEdition)
+        new_edition.more_information = whole_body
+      elsif new_edition.respond_to?(:body)
```
Comment (tadast) [inline app/models/edition.rb]: Minor: `respond_to?(:body=)`? :)

**Comment by tadast** [comment]
May I ask why you chose to check the class in some cases and `respond_to?` in the last case?

File: app/models/edition.rb (hunk 1 above)
Comment (jamiecobbett) [inline app/models/edition.rb]: :+1: rebased

**Comment by binaryberry** [comment]
LGTM! Much cleaner like this.

**Comment by jamiecobbett** [comment]
@tadast you make a very good point - rebased to use `respond_to?`.

In the case of `more_information`, this means that `LocalTransactionEdition` and `PlaceEdition` will now get the `whole_body` from the old edition, which they didn't with my first attempt :+1: 

Yay for duck typing!

**Comment by tadast** [comment]
:cool: I like this change, thanks for addressing the comments :+1:

**Comment by jamiecobbett** [comment]
I've expanded the scope of this a bit to clean up some duplication and some minor bugs. Best reviewed commit-by-commit.

I've tested the changes successfully against Publisher in development.

If people feel this is too much in one PR I can split it out.

**Comment by tadast** [comment]
Other than my comments it looks alright to me, however I'd be more confident if someone with more domain knowledge would have another look :)

**Comment by jamiecobbett** [comment]
@alext has very kindly agreed to review this when he is able to.

**Pull request merged**